.circleci/verbatim-sources/nightly-binary-build-defaults.yml

```diff
@@ -26,70 +26,23 @@
 # (smoke tests and upload jobs do not need the pytorch repo).
 binary_checkout: &binary_checkout
   name: Checkout
-  command: |
-    if [[ -n "$CIRCLE_SHA1" ]]; then
-      # we are on a PR or on a master-merge, but not on a timed build
-      git_url="https://raw.githubusercontent.com/pytorch/pytorch/$CIRCLE_SHA1"
-    else
-      # scheduled nightly binary builds. These run at 05:05 UTC every day. 
-      last_commit="$(git rev-list --before "$(date -u +%Y-%m-%d) 05:00" --max-count 1 HEAD)"
-      git_url="https://raw.githubusercontent.com/pytorch/pytorch/$last_commit"
-    fi
-    retry () {
-        $*  || (sleep 1 && $*) || (sleep 2 && $*) || (sleep 4 && $*) || (sleep 8 && $*)
-    }
-    retry curl -s "$git_url/.circleci/scripts/binary_checkout.sh" -o "binary_checkout.sh"
-    cat "binary_checkout.sh"
-    source "binary_checkout.sh"
+  command: ~/workspace/.circleci/scripts/binary_checkout.sh
 
 # Parses circleci arguments in a consistent way, essentially routing to the
 # correct pythonXgccXcudaXos build we want
 binary_populate_env: &binary_populate_env
   name: Set up env
-  command: |
-    # This step runs on multiple executors with different envfile locations
-    if [[ "$(uname)" == Darwin ]]; then
-      # macos executor (builds and tests)
-      workdir="/Users/distiller/project"
-    elif [[ -d "/home/circleci/project" ]]; then
-      # machine executor (binary tests)
-      workdir="/home/circleci/project"
-    else
-      # docker executor (binary builds)
-      workdir="/"
-    fi
-    script="$workdir/pytorch/.circleci/scripts/binary_populate_env.sh"
-    cat "$script"
-    source "$script"
+  command: ~/workspace/.circleci/scripts/binary_populate_env.sh
```

Comment:
In all the binary jobs, binary_checkout runs first, followed by other binary_ prefixed steps. With your changes, after binary_checkout runs there should be two pytorch checkouts, one at ~/workspace and one at $workdir/pytorch, where $workdir is / on dockers, /home/cirleci/project on machine executors, and /Users/distiller/project on mac executors. I think all the binary_ prefixed steps need to use the $workdir/pytorch checkout instead of the one in ~/workspace.

The binary jobs run at three schedules; there's a few that run on every PR, a few that run on commits merged to master, and a few that run at 9pm PST every day. We need all the binary jobs that run at 9pm every day to use the exact same commit, even if they don't start until a few minutes later. It is not rare for people on the west coast to be landing things at this time, so we can't just pull the latest master in every binary job since it can race. We have this line https://github.com/pytorch/pytorch/blob/master/.circleci/scripts/binary_checkout.sh#L35 in the binary checkout to force the exact same commit across all the binary jobs. This is why we need to use the $workdir/pytorch checkout in the binary jobs.

I think an alternative approach would be to have some parent job that clones master and persists it to the workspace, and then have all the binary jobs require that parent job. But that will take some refactoring.

Reply:
I'll have to digest this a bit more tomorrow (I'm on UK time), but one thing that springs to mind is that we could record the git SHA in the `setup` job, and have every downstream job pick that up.

I'll read over the `binary_` prefxied scripts more carefully - it's worth noting that the checkout in `~/workspace` will only contain the `.circleci/scripts` dir, which should make it fast to share.

Reply:
I've thought about this a little more. I think our options are

1. call binary_checkout from ~/workspace and the rest of the binary_* steps from $workdir. This won't take a lot of work but it is unexpected and'll be confusing to people in the future.
2. Create a binary_setup step, that saves the git SHA of master when it runs, then have all the binary_* jobs require binary_setup. I think this will still require a custom special step in all the binary_* jobs, as it'll have to respect the saved git SHA.
3. Leave the binary jobs as is for now (don't use the new setup step at all in them). Take the huge saving from eliminating setup_ci_env , and look at slimming down the binary_* steps further in a future PR. The binary_* steps are rather slim right now anyways. If we still see problems then I can look into doing #2 later.

Reply:
I've had a look through the `binary_` scripts, they all appear to me to not be location-sensitive, so if they were run from the ~/workspace they would function correctly - is the concern that any changes to these scripts might be inconsistent?

> We need all the binary jobs that run at 9pm every day to use the exact same commit, even if they don't start until a few minutes later. 

I double-checked this with our workflows team, all jobs created as part of a workflow will run with the same CIRCLE_SHA1, so you shouldn't need to do anything extra to acheive this.

I picked a couple of jobs out of the most [recent nightly workflow](https://circleci.com/workflow-run/5d9cc1ab-117b-4e68-a4da-33757160624a)
* https://circleci.com/gh/pytorch/pytorch/1552842
* https://circleci.com/gh/pytorch/pytorch/1553291

You can see from the "Spin up environment" step that these both have `CIRCLE_SHA1` populated to the same ref, and from the output in "Checkout" we can see that the `elif [[ -n "$CIRCLE_SHA1" ]]; then` conditional branch is being used.

Based on my understanding of the scripts I believe that should make the current proposed changes safe - but I could well still be missing something.

-----------

I'm happy to look at implementing option 3 if you'd prefer as that should make the review more straightforward.

Reply:
If that's the case then can you remove the extra Pytorch checkout logic in binary_checkout so that only one pytorch is checked out? Then the rest of these changes should be fine.